Add Cassandra driver Actuator metrics

emerkle826 opened the pull request.

bendiscz committed:
Order metrics auto-configurations correctly

Update metrics auto-configurations so that they are auto-configured
after `CompositeMeterRegistryAutoConfiguration` in order to ensure
the `MeterRegistry` bean has been defined.

Prior to this commit, metrics auto-configurations that depended on a
`MeterRegistry` has `@AutoConfigureAfter(MetricsAutoConfiguration.class)`
which is not sufficient since `MetricsAutoConfiguration` does not export
a `MeterRegistry`.

See gh-21134

adutra commented on spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/cassandra/CassandraMetricsAutoConfiguration.java:
```diff
@@ -0,0 +1,52 @@
+
+@Configuration
+@AutoConfigureAfter({ MetricsAutoConfiguration.class, SimpleMetricsExportAutoConfiguration.class })
+@ConditionalOnClass({ MeterRegistry.class })
+@ConditionalOnBean({ MeterRegistry.class })
```
I confirmed that this autoconfiguration will make the application fail to startup if a `CqlSession` cannot be found:

```
2020-05-22 17:50:37.563 ERROR 25358 --- [           main] o.s.boot.SpringApplication               : Application run failed

java.lang.IllegalStateException: Unable to load cache item
	at org.springframework.cglib.core.internal.LoadingCache.createEntry(LoadingCache.java:79) ~[spring-core-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.cglib.core.internal.LoadingCache.get(LoadingCache.java:34) ~[spring-core-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.cglib.core.AbstractClassGenerator$ClassLoaderData.get(AbstractClassGenerator.java:134) ~[spring-core-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.cglib.core.AbstractClassGenerator.create(AbstractClassGenerator.java:319) ~[spring-core-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.cglib.proxy.Enhancer.createHelper(Enhancer.java:572) ~[spring-core-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.cglib.proxy.Enhancer.createClass(Enhancer.java:419) ~[spring-core-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassEnhancer.createClass(ConfigurationClassEnhancer.java:137) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassEnhancer.enhance(ConfigurationClassEnhancer.java:109) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.enhanceConfigurationClasses(ConfigurationClassPostProcessor.java:423) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanFactory(ConfigurationClassPostProcessor.java:257) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:291) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:131) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:706) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:532) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:143) ~[spring-boot-2.3.0.BUILD-SNAPSHOT.jar:2.3.0.BUILD-SNAPSHOT]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:758) ~[spring-boot-2.3.0.BUILD-SNAPSHOT.jar:2.3.0.BUILD-SNAPSHOT]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:750) ~[spring-boot-2.3.0.BUILD-SNAPSHOT.jar:2.3.0.BUILD-SNAPSHOT]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397) ~[spring-boot-2.3.0.BUILD-SNAPSHOT.jar:2.3.0.BUILD-SNAPSHOT]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:315) ~[spring-boot-2.3.0.BUILD-SNAPSHOT.jar:2.3.0.BUILD-SNAPSHOT]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1237) ~[spring-boot-2.3.0.BUILD-SNAPSHOT.jar:2.3.0.BUILD-SNAPSHOT]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226) ~[spring-boot-2.3.0.BUILD-SNAPSHOT.jar:2.3.0.BUILD-SNAPSHOT]
	at com.example.demo.DemoApplication.main(DemoApplication.java:10) ~[classes/:na]
Caused by: java.lang.NoClassDefFoundError: com/datastax/oss/driver/api/core/CqlSession
	at java.base/java.lang.Class.getDeclaredMethods0(Native Method) ~[na:na]
	at java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3172) ~[na:na]
	at java.base/java.lang.Class.getDeclaredMethods(Class.java:2315) ~[na:na]
	at org.springframework.cglib.core.ReflectUtils.addAllMethods(ReflectUtils.java:451) ~[spring-core-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.cglib.proxy.Enhancer.getMethods(Enhancer.java:635) ~[spring-core-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.cglib.proxy.Enhancer.generateClass(Enhancer.java:670) ~[spring-core-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.cglib.transform.TransformingClassGenerator.generateClass(TransformingClassGenerator.java:33) ~[spring-core-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.cglib.core.DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25) ~[spring-core-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.cglib.core.ClassLoaderAwareGeneratorStrategy.generate(ClassLoaderAwareGeneratorStrategy.java:57) ~[spring-core-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.cglib.core.AbstractClassGenerator.generate(AbstractClassGenerator.java:358) ~[spring-core-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.cglib.proxy.Enhancer.generate(Enhancer.java:585) ~[spring-core-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.cglib.core.AbstractClassGenerator$ClassLoaderData$3.apply(AbstractClassGenerator.java:110) ~[spring-core-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.cglib.core.AbstractClassGenerator$ClassLoaderData$3.apply(AbstractClassGenerator.java:108) ~[spring-core-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.cglib.core.internal.LoadingCache$2.call(LoadingCache.java:54) ~[spring-core-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
	at org.springframework.cglib.core.internal.LoadingCache.createEntry(LoadingCache.java:61) ~[spring-core-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	... 21 common frames omitted
Caused by: java.lang.ClassNotFoundException: com.datastax.oss.driver.api.core.CqlSession
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:583) ~[na:na]
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178) ~[na:na]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521) ~[na:na]
	... 37 common frames omitted
```

```suggestion
@ConditionalOnClass({ MeterRegistry.class, CqlSession.class })
@ConditionalOnBean({ MeterRegistry.class, CqlSession.class })
```

tomekl007 replied:
+1 on using `ConditionalOnClass` - it checks the classpath and doesn't fail.
We should add a test that detects that and validate the behavior when the java-driver is not on the classpath.
You can use a similar mechanism to this:
https://github.com/riptano/java-driver-spring/blob/master/starter/src/test/java/com/datastax/oss/spring/configuration/metrics/DataStaxDriverMetricsAutoConfigurationTest.java#L33-L40
mainly the `new FilteredClassLoader(MeterRegistry.class)` that allows you to filter out classes from the classpath

emerkle826 replied:
I've added a test [here](https://github.com/datastax/spring-boot/pull/2/files#diff-8991f02ba760e2c1590560601a62f5a1R118-R132) for this.

adutra commented on spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/cassandra/CassandraMetricsAutoConfiguration.java:
```diff
@@ -0,0 +1,52 @@
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@AutoConfigureAfter({ MetricsAutoConfiguration.class, SimpleMetricsExportAutoConfiguration.class })
```
Seems like we need to defer configuration of metrics until `CassandraAutoConfiguration` has produced a `CqlSession` bean (or the user has produced one in lieu of the default one):

```suggestion
@AutoConfigureAfter({ MetricsAutoConfiguration.class, SimpleMetricsExportAutoConfiguration.class,
		CassandraAutoConfiguration.class })
```

adutra commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/cassandra/CassandraMetricsSettings.java:
```diff
@@ -0,0 +1,50 @@
+ * @author Erik Merkle
+ * @since 2.3.0
+ */
+@ConfigurationProperties(prefix = "cassandra.driver.session.metrics")
```
The existing `CassandraProperties`  class has prefix `spring.data.cassandra`. I think we need to keep our stuff inside that prefix, so I'd suggest `spring.data.cassandra.metrics` – hoping that this is not going to create a confusion when loading properties. If it does, maybe use `spring.data.metrics.cassandra` instead.

adutra commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/cassandra/CassandraMetricsBinder.java:
```diff
@@ -0,0 +1,153 @@
+				registerMeter(name, (Counting) m, meterRegistry);
+			}
+			else if (m instanceof com.codahale.metrics.Gauge) {
+				registerGauge(name, (com.codahale.metrics.Gauge) m, meterRegistry);
```
```suggestion
				registerGauge(name, (com.codahale.metrics.Gauge<?>) m, meterRegistry);
```

adutra commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/cassandra/CassandraMetricsBinder.java:
```diff
@@ -0,0 +1,153 @@
+
+	private final CqlSession session;
+
+	private CassandraMetricsSettings cassandraMetricsSettings;
```
```suggestion
	private final CassandraMetricsSettings cassandraMetricsSettings;
```

adutra commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/cassandra/CassandraMetricsBinder.java:
```diff
@@ -0,0 +1,153 @@
+			@NonNull CassandraMetricsSettings cassandraMetricsSettings) {
+		this.cassandraMetricsSettings = cassandraMetricsSettings;
+		this.session = cqlSession;
+		this.tags = Tags.concat(Collections.emptyList(), "sessionName", cqlSession.getName());
```
```suggestion
		this.tags = Tags.of("sessionName", cqlSession.getName());
```

adutra commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/cassandra/CassandraMetricsBinder.java:
```diff
@@ -0,0 +1,153 @@
+				.register(meterRegistry);
+	}
+
+	private void registerGauge(String metricName, com.codahale.metrics.Gauge gauge, MeterRegistry meterRegistry) {
```
```suggestion
	private void registerGauge(String metricName, com.codahale.metrics.Gauge<?> gauge, MeterRegistry meterRegistry) {
```

adutra commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/cassandra/CassandraMetricsBinder.java:
```diff
@@ -0,0 +1,153 @@
+							(t) -> convertToTimeUnit(supplier.apply(timer.getSnapshot()).doubleValue()))
+					.tags(this.tags).baseUnit(this.cassandraMetricsSettings.getTimerTimeUnit().name())
+					.description(DESCRIPTION).register(meterRegistry);
+		});
```
You should create the snapshot only once:

```suggestion
		Snapshot snapshot = timer.getSnapshot();
		TIMER_SNAPSHOT_METRICS.forEach((n, supplier) -> {
			String nameTimerMetric = String.format("%s.%s", name, n);
			FunctionCounter
					.builder(nameTimerMetric, timer, (t) -> convertToTimeUnit(supplier.apply(snapshot).doubleValue()))
					.tags(this.tags).baseUnit(this.cassandraMetricsSettings.getTimerTimeUnit().name())
					.description(DESCRIPTION).register(meterRegistry);
		});
```

adutra commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/cassandra/CassandraMetricsBinder.java:
```diff
@@ -0,0 +1,153 @@
+		TIMER_METRICS.forEach((n, supplier) -> {
+			String nameTimerMetric = String.format("%s.%s", name, n);
+			FunctionCounter.builder(nameTimerMetric, timer, (t) -> supplier.apply(timer).doubleValue()).tags(this.tags)
+					.baseUnit(TimeUnit.SECONDS.name()) // all rates are calculated per
+					// second
```
```suggestion
					.baseUnit(TimeUnit.SECONDS.name()) // all rates are calculated per second
```

emerkle826 replied:
The gradle format task actually does this to the comment. I'll fix it so that it's compliant with the format check.

adutra commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/cassandra/CassandraMetricsBinder.java:
```diff
@@ -0,0 +1,153 @@
+	/**
+	 * Description for micrometer metrics.
+	 */
+	public static final String DESCRIPTION = "Please see reference.conf for more information.";
```
```suggestion
	private static final String DESCRIPTION = "Please see reference.conf for more information.";
```

adutra commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/cassandra/CassandraMetricsBinder.java:
```diff
@@ -0,0 +1,153 @@
+		TIMER_METRICS.put("m1_rate", Timer::getOneMinuteRate);
+		TIMER_METRICS.put("m5_rate", Timer::getFiveMinuteRate);
+		TIMER_METRICS.put("m15_rate", Timer::getFifteenMinuteRate);
+	}
```
I'd rather wrap these map in an unmodifiable map:

```suggestion
	private static final Map<String, Function<Snapshot, Number>> TIMER_SNAPSHOT_METRICS;

	private static final Map<String, Function<Timer, Number>> TIMER_METRICS;

	static {
		Map<String, Function<Snapshot, Number>> map = new HashMap<>();
		map.put("p999", Snapshot::get999thPercentile);
		map.put("p99", Snapshot::get99thPercentile);
		map.put("p98", Snapshot::get98thPercentile);
		map.put("p95", Snapshot::get95thPercentile);
		map.put("p75", Snapshot::get75thPercentile);
		map.put("min", Snapshot::getMin);
		map.put("max", Snapshot::getMax);
		map.put("mean", Snapshot::getMean);
		map.put("median", Snapshot::getMedian);
		map.put("stdDev", Snapshot::getStdDev);
		TIMER_SNAPSHOT_METRICS = Collections.unmodifiableMap(map);
	}

	static {
		Map<String, Function<Timer, Number>> map = new HashMap<>();
		map.put("mean_rate", Timer::getMeanRate);
		map.put("m1_rate", Timer::getOneMinuteRate);
		map.put("m5_rate", Timer::getFiveMinuteRate);
		map.put("m15_rate", Timer::getFifteenMinuteRate);
		TIMER_METRICS = Collections.unmodifiableMap(map);
	}
```

adutra commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/cassandra/CassandraMetricsBinder.java:
```diff
@@ -0,0 +1,153 @@
+			FunctionCounter
+					.builder(nameTimerMetric, timer,
+							(t) -> convertToTimeUnit(supplier.apply(timer.getSnapshot()).doubleValue()))
+					.tags(this.tags).baseUnit(this.cassandraMetricsSettings.getTimerTimeUnit().name())
```
I'm struggling to follow. You are converting the measures to nanoseconds, but reporting the measure to be in the original unit, which doesn't make sense to me.

emerkle826 replied:
I've removed this conversion, and the CassandraMetricsSettings class in general.

tomekl007 replied:
Why do you want to remove it? Don't you think that converting may be useful for end-users? 
(I don't know if spring allows that for other modules)

adutra replied:
Honestly nanoseconds is the best you can do.

emerkle826 replied:
The reason for the conversion was to allow users to control the granularity of the timer snapshot. Nanoseconds is the finest we can go, but maybe that is a bit overkill for some situations (thinking about web apps in general here).

Initially, we had another attribute in the settings class that would allow you to disable metrics, even if some of them were enabled in the driver config. This is not necessary if we are contributing metrics to Spring Boot as we can disable them via the Spring metrics management settings (built-in). So that leaves just this one config attribute in CassandraMetricsSettings.

So the question is, is having a configurable time unit for these metrics worth the extra class and setting? If so, I can bring it back, but I will have to resume searching for what documentation needs to go where for the setting as one of the gradle documentation tasks (`documentConfigurationProperties`) fails for all the attributes in CassandraMetricsSettings.

emerkle826 replied:
I think it will be easier to just go with nanosecond granularity here. I don't know that there is a lot of value in trying to give users the flexibility to control it.

adutra commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/cassandra/CassandraMetricsSettings.java:
```diff
@@ -0,0 +1,50 @@
+ * @since 2.3.0
+ */
+@ConfigurationProperties(prefix = "cassandra.driver.session.metrics")
+public class CassandraMetricsSettings {
```
In fact, do we really need `CassandraMetricsSettings` at all? Why can't we simply expose the timer snapshot measures in milliseconds?

emerkle826 replied:
I'll remove this class as it will make the TimerUnit conversion above moot as well as simplify things.

adutra commented on spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/cassandra/CassandraMetricsAutoConfiguration.java:
```diff
@@ -0,0 +1,52 @@
+		if (!cqlSession.getMetrics().isPresent()) {
+			return;
+		}
+		cassandraMetrics.bindTo(registry);
```
```suggestion
		// Don't bind to the MeterRegistry if the session has no metrics configured
		if (!cqlSession.getMetrics().isPresent()) {
			return;
		}
		CassandraMetricsBinder cassandraMetrics = new CassandraMetricsBinder(cqlSession, cassandraMetricsSettings);
		cassandraMetrics.bindTo(registry);
```

adutra commented on spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/metrics/cassandra/CassandraMetricsAutoConfigurationTests.java:
```diff
@@ -0,0 +1,108 @@
+ *
+ * @author Erik Merkle
+ */
+public class CassandraMetricsAutoConfigurationTests {
```
These tests will only work if there is a Cassandra node listening on localhost:9042. You need to mock the session instead:

```
public class CassandraMetricsAutoConfigurationTests {

	private static final String METRIC_NAME = CassandraMetricsBinder.SESSION_PREFIX + ".s0.bytes-sent";

	private ApplicationContextRunner contextRunner = new ApplicationContextRunner().with(MetricsRun.simple())
			.withConfiguration(
					AutoConfigurations.of(CassandraAutoConfiguration.class, CassandraMetricsAutoConfiguration.class));

	/**
	 * Cassandra driver metrics should be enabled by default as long as the desired
	 * metrics are enabled in the Driver's configuration.
	 */
	@Test
	void autoConfiguredCassandraIsInstrumented() {
		this.contextRunner
				.withBean(CqlSession.class, () -> mockSession(true))
				.run((context) -> {
					MeterRegistry registry = context.getBean(MeterRegistry.class);
					Meter meter = registry.find(METRIC_NAME).meter();
					assertThat(meter).isNotNull();
					assertThat(meter.measure()).hasSize(1).extracting("value").containsExactly(42d);
				});
	}

	/**
	 * Cassandra driver metrics should be disabled if set to false via Spring metrics
	 * management.
	 */
	@Test
	void cassandraInstrumentationCanBeDisabledThroughSpringMetricsManagement() {
		this.contextRunner
				.withBean(CqlSession.class, () -> mockSession(true))
				.withPropertyValues("management.metrics.enable.cassandra.driver.session=false")
				.run((context) -> {
					MeterRegistry registry = context.getBean(MeterRegistry.class);
					assertThat(registry.find(METRIC_NAME).meter()).isNull();
				});
	}

	/**
	 * Cassandra driver metrics should be disabled if no metrics are enabled in the
	 * Driver's configuration.
	 */
	@Test
	void cassandraInstrumentationIsDisabledIfNoMetricsEnabledInDriverConfig() {
		this.contextRunner
				.withBean(CqlSession.class, () -> mockSession(false))
				.run((context) -> {
					MeterRegistry registry = context.getBean(MeterRegistry.class);
					assertThat(registry.find(METRIC_NAME).meter()).isNull();
				});
	}

	private CqlSession mockSession(boolean enableMetrics) {
		CqlSession session = mock(CqlSession.class);
		given(session.getName()).willReturn("s0");
		if (enableMetrics) {
			Metrics metrics = mock(Metrics.class);
			given(session.getMetrics()).willReturn(Optional.of(metrics));
			MetricRegistry metricRegistry = new MetricRegistry();
			given(metrics.getRegistry()).willReturn(metricRegistry);
			metricRegistry.meter("s0.bytes-sent").mark(42);
		}
		else {
			given(session.getMetrics()).willReturn(Optional.empty());
		}
		return session;
	}
}
```

emerkle826 replied:
You're right, I missed this when testing and didn't realize I had a CCM instance up.

tomekl007 commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/cassandra/CassandraMetricsBinder.java:
```diff
@@ -0,0 +1,153 @@
+		String name = String.format("%s.%s", SESSION_PREFIX, metricName);
+
+		// count metric should be registered at the root level of the metric name
+		FunctionCounter.builder(name, timer, Timer::getCount).tags(this.tags).description(DESCRIPTION)
```
nit: you don't need `this.`

emerkle826 replied:
The Spring checkStyle settings require `this.` for all local attribute references.

tomekl007 commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/cassandra/CassandraMetricsBinder.java:
```diff
@@ -0,0 +1,153 @@
+
+		TIMER_METRICS.forEach((n, supplier) -> {
+			String nameTimerMetric = String.format("%s.%s", name, n);
+			FunctionCounter.builder(nameTimerMetric, timer, (t) -> supplier.apply(timer).doubleValue()).tags(this.tags)
```
nit: you don't need `this.`

emerkle826 replied:
again, this is a Spring checkStyle thing.

tomekl007 commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/cassandra/CassandraMetricsBinder.java:
```diff
@@ -0,0 +1,153 @@
+	/**
+	 * Prefix for all Driver configuration parameters.
+	 */
+	public static final String SESSION_PREFIX = "cassandra.driver.session";
```
shouldn't it be `spring.data.cassandra.driver.session`? (to comply with the `spring.data.cassandra` prefix convention)

tomekl007 commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/cassandra/CassandraMetricsBinder.java:
```diff
@@ -0,0 +1,153 @@
+		});
+	}
+
+	private long convertToTimeUnit(double value) {
```
You should create a unit or IT tests that are covering conversion of time, similar to:
https://github.com/riptano/java-driver-spring/blob/master/starter/src/test/java/com/datastax/oss/spring/configuration/metrics/DriverMetricsTest.java#L96-L158

emerkle826 replied:
I will, if we decide to bring the config setting back.

tomekl007 commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/cassandra/CassandraMetricsBinder.java:
```diff
@@ -0,0 +1,153 @@
+	public void bindTo(@NonNull MeterRegistry meterRegistry) {
+		Assert.isTrue(this.session.getMetrics().isPresent(), "Metrics on the CqlSession needs to be present");
+
+		this.session.getMetrics().get().getRegistry().getMetrics().forEach((name, m) -> {
```
to validate if we are correctly registering different types of metrics, we should create an IT test that checks at least one metric of every type. Maybe you could follow this test:
https://github.com/riptano/java-driver-spring/blob/master/integration-tests/src/test/java/com/datastax/oss/spring/configuration/exampletests/CassandraMetricsMicrometerIT.java
It iterates over all metrics and validates that they are registered properly. I was able to catch some bugs when implementing original version using this test.

emerkle826 replied:
I've added some more example metrics to the test [here](https://github.com/datastax/spring-boot/pull/2/files#diff-8991f02ba760e2c1590560601a62f5a1R58-R83).

Larisho committed:
Migrate legacy endpoint that have a dash in their id

See gh-21615

emerkle826 commented:
@adutra and @tomekl007 I've added a few more tests and I think I've addressed the feedback. I'd appreciate one last quick check when you have a chance and then I'll move the PR to Spring's repo.

tomekl007 commented on spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/metrics/cassandra/CassandraMetricsAutoConfigurationTests.java:
```diff
@@ -0,0 +1,168 @@
+	 * on the classpath.
+	 */
+	@Test
+	void cassandraInstrumentationIsDisabledIfDriverClassesNotOnClasspath() {
```
I think that we should also test the case when `MetricsRegistry` AND `CqlSession` is not on the classpath. In that case the `MeterRegistry.getBean` should return null.

emerkle826 replied:
It doesn't make sense to try to test with `CqlSession` filtered from the classpath. The ApplicationContextRunner tries to run with `CassandraMetricsAutoConfiguration`, which requires a `CqlSession` bean. If `CqlSession` isn't on the classpath, trying to spin up a context with `CassandraMetricsAutoConfiguration` will throw an exception indicating it can't find a required bean.

I can change the test so that it removes `MetricRegistry` from the classpath, but I'm mocking the `CqlSession` bean with a `MetricRegistry` since the tests don't have a fully initialized `CqlSession` bean with metrics configuration. I don't think removing `MetricRegistry` from the classpath at this point will prevent the metrics from being bound to the Micrometer `MeterRegistry`.

Or am I missing something?

adutra replied:
> It doesn't make sense to try to test with CqlSession filtered from the classpath. The ApplicationContextRunner tries to run with CassandraMetricsAutoConfiguration, which requires a CqlSession bean. If CqlSession isn't on the classpath, trying to spin up a context with CassandraMetricsAutoConfiguration will throw an exception indicating it can't find a required bean.

It shouldn't throw an exception, it should instead not be executed as part of the configuration process, since that class has a `@ConditionalOnClass({ MeterRegistry.class, CqlSession.class })` annotation.

This specific test does exactly that: it includes `CassandraMetricsAutoConfiguration` in the context, but excludes `CqlSession` from the classpath, to emulate the situation where the driver is not on the classpath.

I think what @tomekl007 wants is just another test similar to this one, but in which `MeterRegistry` would also be absent from the classpath, to emulate the situation where the user has excluded Micrometer _and_ the driver from the classpath.

I personally don't think that this is a common situation, but imo it doesn't hurt to add a test for that.

emerkle826 replied:
I understand the intent. But I'm seeing the exception when I run the tests if I don't provide a CqlSession bean to the test's ApplicationContextRunner, i.e.

```java
this.contextRunner.withBean(CqlSession.class, () -> mockSession(true))
```
If I omit the bean, I get a NoSuchBeanDefinitionException with

```
No qualifying bean of type 'com.datastax.oss.driver.api.core.CqlSession' available.
```
Does it make sense to create the context with the mocked bean, but test without `CqlSession` on the classpath?

emerkle826 replied:
I think I see the problem....

emerkle826 replied:
Ok, with the latest changes, I can spin up the context with `CqlSession` and `MetricRegistry` missing from the classpath. The Micrometer metrics are indeed disabled, however the Micrometer `MeterRegisrty` is not null, it simply has no meters in its map.

tomekl007 commented on spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/metrics/cassandra/CassandraMetricsAutoConfigurationTests.java:
```diff
@@ -0,0 +1,168 @@
+	void autoConfiguredCassandraIsInstrumented() {
+		this.contextRunner.withBean(CqlSession.class, () -> mockSession(true)).run((context) -> {
+			MeterRegistry registry = context.getBean(MeterRegistry.class);
+			// Assert a typical Driver Meter metric
```
nit: can we extract assertions of metrics to separate methods for readability? i.e:
https://github.com/datastax/spring-boot/pull/2/files#diff-8991f02ba760e2c1590560601a62f5a1R59-R61
to `assertMeterMetric()`, then `assertGaugeMetric()`, etc

adutra commented:
Official PR opened here: https://github.com/spring-projects/spring-boot/pull/21624.

AndrewDi committed:
Add basic auth support for Prometheus pushgateway

See gh-22548

anshlykov committed:
Add pullPolicy option for image building

This commit adds a pullPolicy option to the configuration of the Maven
plugin spring-boot:build-image goal and the Gradle plugin bootBuildImage
task. The new option gives users control over pulling the builder image
and run image from a remote image registry to the local Docker daemon.

See gh-22736

dreis2211 committed:
Introduce @ForkedClassPath for testing unmodified class path

See gh-22710

Lopfest committed:
Fix include exception handling in DefaultErrorAttributes

This commit fixes a problem with the handling of the includeException
field in DefaultErrorAttributes.

See gh-22750

Commit:
Fix wrong reference in `ReadinessStateHealthIndicator` javadoc

See gh-22760

izeye committed:
Mention configuration customizers for WebTestClient and RestAssured

See gh-22765

mbhave committed:
Do not ignore null/empty values from SPRING_APPLICATION_JSON

Fixes gh-21542

davidbilge committed:
Propagate the cause that led Reactor"s debug agent to fail

Prior to this commit, DebugAgentEnvironmentPostProcessor throws a
RuntimeException with a generic error message if its initialization
fails. The causing exception is discarded, which makes error analysis
unnecessarily difficult.

This commit attaches the cause to the newly thrown RuntimeException.

See gh-22777

wilkinsona committed:
Upgrade to Build Helper Maven Plugin 3.2.0

Closes gh-22788

wilkinsona committed:
Upgrade to Byte Buddy 1.10.14

Closes gh-22789

wilkinsona committed:
Upgrade to Caffeine 2.8.5

Closes gh-22790

wilkinsona committed:
Upgrade to Cassandra Driver 4.8.0

Closes gh-22791

wilkinsona committed:
Upgrade to Commons Lang3 3.11

Closes gh-22792

wilkinsona committed:
Upgrade to Commons Pool2 2.8.1

Closes gh-22793

wilkinsona committed:
Upgrade to DB2 JDBC 11.5.4.0

Closes gh-22794

wilkinsona committed:
Upgrade to Dropwizard Metrics 4.1.11

Closes gh-22795

wilkinsona committed:
Upgrade to Flyway 6.5.3

Closes gh-22796

wilkinsona committed:
Upgrade to Hibernate 5.4.19.Final

Closes gh-22797

wilkinsona committed:
Upgrade to HSQLDB 2.5.1

Closes gh-22798

wilkinsona committed:
Upgrade to HtmlUnit 2.42.0

Closes gh-22799

wilkinsona committed:
Upgrade to Infinispan 11.0.3.Final

Closes gh-22800

wilkinsona committed:
Upgrade to Jackson Bom 2.11.2

Closes gh-22801

wilkinsona committed:
Upgrade to Jetty 9.4.31.v20200723

Closes gh-22802

wilkinsona committed:
Upgrade to Johnzon 1.2.8

Closes gh-22803

wilkinsona committed:
Upgrade to jOOQ 3.13.4

Closes gh-22804

wilkinsona committed:
Upgrade to Lettuce 5.3.2.RELEASE

Closes gh-22805

wilkinsona committed:
Upgrade to Liquibase 3.10.2

Closes gh-22806

wilkinsona committed:
Upgrade to MariaDB 2.6.2

Closes gh-22807

wilkinsona committed:
Upgrade to Maven War Plugin 3.3.1

Closes gh-22808

wilkinsona committed:
Upgrade to Micrometer 1.5.3

Closes gh-22809

wilkinsona committed:
Upgrade to MongoDB 4.1.0

Closes gh-22810

wilkinsona committed:
Upgrade to MySQL 8.0.21

Closes gh-22811

wilkinsona committed:
Upgrade to Netty 4.1.51.Final

Closes gh-22812

wilkinsona committed:
Upgrade to Selenium HtmlUnit 2.42.0

Closes gh-22813

wilkinsona committed:
Upgrade to SendGrid 4.6.3

Closes gh-22814

wilkinsona committed:
Upgrade to Solr 8.5.2

Closes gh-22815

wilkinsona committed:
Upgrade to Spring HATEOAS 1.1.1.RELEASE

Closes gh-22816

wilkinsona committed:
Upgrade to SQLite JDBC 3.32.3.2

Closes gh-22817

wilkinsona committed:
Upgrade to Tomcat 9.0.37

Closes gh-22818

Commit:
Exclude cookie headers by default from HTTP traces

See gh-22829

wilkinsona committed:
Configure generated snippets as inputs to the Asciidoctor tasks

Closes gh-22819

wilkinsona committed:
Merge branch '2.3.x'

Closes gh-22820

bclozel committed:
Update HTTP/2 auto-configuration for Jetty

Prior to this commit, the HTTP/2 server auto-configuration for Jetty
would require Conscrypt as a hard dependency.

This commit updates the auto-configuration for more flexibility and now
allows the following deployments:

* JDK9+ with the JDK ALPN implementation
* JDK8u252+ with the backported ALPN implementation
* Conscrypt with no JDK requirement

The auto-configuration now improves detection and guides developers in
case there is a missing `jetty-alpn-*-server` dependency.

The reference docs in the HOWTO section has been updated accordingly.

Closes gh-22188

bclozel committed:
Merge branch '2.3.x'

Closes gh-22826

bclozel committed:
End StartupSteps only once in SpringApplicationRunListeners

Fixes gh-22776

wilkinsona committed:
Isolate Maven Plugin's integration tests from repo.spring.io

Previously, the Maven plugin integration tests used a settings.xml file
that defined https://repo.spring.io/snapshot as a repository. This
allowed them to resolve snapshots of the plugin's Spring Framework
dependencies but it had the unfortunate side-effect of also allowing
them to resolve snapshots of other Spring Boot modules from Artifactory
rather than using those currently being built.

This commit replaces the repositories in settings.xml with a Gradle
task that resolves the necessary dependencies and populates a local
repository with the dependencies' jars and pom files. This is achieved
using a ComponentMetadataRule that creates a custom variant of each
dependency that includes its pom file, inspired by the example in
gradle/gradle/#11449. A configuration that extends the
runtimeClasspath configuration and select the custom variant via its
attribute is then used to resolve the jars and pom files of the runtime
classpath such that they can then be used to populate the local
repository.

Closes gh-22828

wilkinsona committed:
Merge branch '2.3.x'

Closes gh-22831

bclozel committed:
Adapt to MonoProcessor deprecations in Reactor

This commit updates our usage of `MonoProcessor` after the deprecations
introduced in reactor/reactor-core#1053

wilkinsona committed:
Align with changes in latest Integration snapshots

See gh-22729

scottfrederick committed:
Preserve user-provided values in MongoClientSettings

Prior to this commit, values from MongoProperties would always
overwrite matching fields in MongoClientSettings. This commit
preserves all values in MongoClientSettings if the client app
provides the MongoClientSettings bean, and only overwrites from
MongoProperties if no MongoClientSettings bean is provided.

Fixes gh-22321

wilkinsona committed:
Upgrade to Mockito 3.4.6

Closes gh-22838

dreis2211 committed:
Fix some doc anchors

See gh-22839

snicoll committed:
Start building against Reactor Dysprosium-SR11 snapshots

See gh-22841

snicoll committed:
Start building against Spring AMQP 2.2.10 snapshotsEnabled

See gh-22842

snicoll committed:
Start building against Spring Kafka 2.5.5 snapshots

See gh-22843

snicoll committed:
Start building against Spring Data Neumann SR3 snapshots

See gh-22844

snicoll committed:
Merge branch '2.3.x'

mdeinum committed:
Allow step action to be optional

Update `SpringApplicationRunListeners` so that the step action
is optional and does not need to call `end()`.

This commit also introduces a test for the failure case showing
that there is a missed call to end with the initial solution.

See gh-22845

snicoll committed:
Merge pull request #22760 from ksmilek-wr

* pr/22760:
  Fix wrong reference in `ReadinessStateHealthIndicator` javadoc

Closes gh-22760

snicoll committed:
Merge branch '2.3.x'

Closes gh-22846

snicoll committed:
Upgrade to Reactor Dysprosium-SR11

Closes gh-22841

snicoll committed:
Merge branch '2.3.x'

snicoll committed:
Polish "Propagate the cause that led Reactor"s debug agent to fail"

See gh-22777

snicoll committed:
Merge pull request #22777 from davidbilge

* pr/22777:
  Polish "Propagate the cause that led Reactor"s debug agent to fail"
  Propagate the cause that led Reactor"s debug agent to fail

Closes gh-22777

snicoll committed:
Merge branch '2.2.x' into 2.3.x

Closes gh-22847

snicoll committed:
Merge branch '2.3.x'

Closes gh-22848

snicoll committed:
Polish

snicoll committed:
Merge branch '2.2.x' into 2.3.x

snicoll committed:
Merge branch '2.3.x'

snicoll committed:
Merge pull request #22839 from dreis2211

* pr/22839:
  Fix some doc anchors

Closes gh-22839

snicoll committed:
Polish "Migrate legacy endpoint that have a dash in their id"

See gh-21615

snicoll committed:
Merge pull request #21615 from Larisho

* pr/21615:
  Polish "Migrate legacy endpoint that have a dash in their id"
  Migrate legacy endpoint that have a dash in their id

Closes gh-21615

snicoll committed:
Merge branch '2.2.x' into 2.3.x

Closes gh-22849

snicoll committed:
Merge branch '2.3.x'

Closes gh-22850

snicoll committed:
Upgrade to Reactor 2020.0.0-M2

Closes gh-22667

snicoll committed:
Upgrade to RSocket 1.1.0-M1

Closes gh-22751

wilkinsona committed:
Only enable full path optimization when there's one DispatcherServlet

Previously, UrlPathHelper's full path optimization was enabled when
there was a dispatcher servlet mapped to /. The UrlPathHelper is used
across Spring MVC and if there are multiple dispatcher servlets they
all share the sample UrlPathHelper. This meant that any additional
dispatcher servlets mapping to locations other than / would not be able
to map requests correctly as the UrlPathHelper would use the full path,
ignoring the url mapping of the dispatcher servlet.

This commit updates the MVC auto-configuration so that use of the full
path is only enabled if there's a single dispatcher servlet
registration.

Fixes gh-22682

snicoll committed:
Upgrade to Spring Framework 5.3.0-M2

Closes gh-22640

wilkinsona committed:
Merge branch '2.3.x'

Closes gh-22856

wilkinsona committed:
Upgrade to Dependency Management Plugin 1.0.10.RELEASE

Closes gh-22615

Commit:
Support ConfigurationClassPostProcessor supplier

Update `SharedMetadataReaderFactoryContextInitializer` to optionally
support a `ConfigurationClassPostProcessor` provided via an instance
`Supplier`.

If the instance supplier is available then we can use that to set the
`MetadataReaderFactory` non-reflectively, instead of using a
`PropertyValue` (and the `BeanWrapper` eventually).

The previous behavior is left as a  fallback and without a change in
Spring Framework will continue to be the default for most applications.

See gh-22858

wilkinsona committed:
Merge branch '2.2.x' into 2.3.x

wilkinsona committed:
Merge branch '2.3.x'

wilkinsona committed:
Upgrade to Byte Buddy 1.10.14

Closes gh-22859

wilkinsona committed:
Upgrade to Commons Pool2 2.8.1

Closes gh-22860

wilkinsona committed:
Upgrade to Couchbase Client 3.0.7

Closes gh-22861

wilkinsona committed:
Upgrade to Dependency Management Plugin 1.0.10.RELEASE

Closes gh-22862

wilkinsona committed:
Upgrade to Dropwizard Metrics 4.1.12.1

Closes gh-22863

wilkinsona committed:
Upgrade to Hibernate 5.4.20.Final

Closes gh-22864

wilkinsona committed:
Upgrade to Jackson Bom 2.11.2

Closes gh-22865

wilkinsona committed:
Upgrade to Jetty 9.4.31.v20200723

Closes gh-22866

wilkinsona committed:
Upgrade to jOOQ 3.13.4

Closes gh-22867

wilkinsona committed:
Upgrade to Kafka 2.5.1

Closes gh-22868

wilkinsona committed:
Upgrade to Lettuce 5.3.3.RELEASE

Closes gh-22869

wilkinsona committed:
Upgrade to Pooled JMS 1.1.2

Closes gh-22870

wilkinsona committed:
Upgrade to RSocket 1.0.2

Closes gh-22871

wilkinsona committed:
Upgrade to Spring HATEOAS 1.1.1.RELEASE

Closes gh-22872

wilkinsona committed:
Upgrade to Spring Security 5.3.4.RELEASE

Closes gh-22873

snicoll committed:
Restore customization of the Couchbase cache manager

With the upgrade to the new Couchbase SDK and the related changes in
Spring Data Couchbase, CacheManagerCustomizer can no longer be used to
customize the Couchbase cache manager as it is an immutable class.

This commit introduces a dedicated callback for the
CouchbaseCacheManagerBuilder that is used by the auto-configuration and
update the documentation to refer to it with a sample usage.

Closes gh-22573

snicoll committed:
Merge branch '2.3.x'

Closes gh-22874

wilkinsona committed:
Merge branch '2.3.x'

wilkinsona committed:
Upgrade to Dependency Management Plugin 1.0.10.RELEASE

Closes gh-22875

wilkinsona committed:
Upgrade to Dropwizard Metrics 4.1.12.1

Closes gh-22876

wilkinsona committed:
Upgrade to Hibernate 5.4.20.Final

Closes gh-22877

wilkinsona committed:
Upgrade to HtmlUnit 2.43.0

Closes gh-22878

wilkinsona committed:
Upgrade to Kotlin Coroutines 1.3.8

Closes gh-22879

wilkinsona committed:
Upgrade to Lettuce 5.3.3.RELEASE

Closes gh-22880

wilkinsona committed:
Upgrade to Maven Resources Plugin 3.2.0

Closes gh-22881

wilkinsona committed:
Upgrade to Pooled JMS 1.1.2

Closes gh-22882

wilkinsona committed:
Upgrade to Selenium HtmlUnit 2.43.0

Closes gh-22883

wilkinsona committed:
Upgrade to Versions Maven Plugin 2.8.1

Closes gh-22884

wilkinsona committed:
Avoid duplicates with different versions in Devtools' int test deps

Closes gh-22887

wilkinsona committed:
Merge branch '2.3.x'

Closes gh-22888

snicoll committed:
Document that Cassandra's application.conf is not considered

Closes gh-21787

snicoll committed:
Merge branch '2.3.x'

Closes gh-22889

wilkinsona committed:
Polish "Exclude cookie headers by default from HTTP traces"

See gh-22829

snicoll committed:
Polish

snicoll committed:
Document the limitation of default value detection

Closes gh-22685

snicoll committed:
Merge branch '2.2.x' into 2.3.x

Closes gh-22890

snicoll committed:
Merge branch '2.3.x'

Closes gh-22891

wilkinsona committed:
Merge pull request #22829 from emilytsanova

* gh-22829:
  Polish "Exclude cookie headers by default from HTTP traces"
  Exclude cookie headers by default from HTTP traces

Closes gh-22829

wilkinsona committed:
Upgrade to Spring HATEOAS 1.2.0-M1

Closes gh-22893

philwebb committed:
Polish 'Allow step action to be optional'

See gh-22845

philwebb committed:
Merge pull request #22845 from mdeinum

* pr/22845:
  Polish 'Allow step action to be optional'
  Allow step action to be optional

Closes gh-22845

philwebb committed:
Polish 'Support ConfigurationClassPostProcessor supplier'

See gh-22858

mbhave committed:
Merge branch '2.2.x' into 2.3.x

Closes gh-22895

mbhave committed:
Merge branch '2.3.x'

Closes gh-22897

philwebb committed:
Merge pull request #22858 from dsyer

* pr/22858:
  Polish 'Support ConfigurationClassPostProcessor supplier'
  Support ConfigurationClassPostProcessor supplier

Closes gh-22858

scottfrederick committed:
Polish "Add pullPolicy option for image building"

See gh-22736

johncfranco committed:
Harmonize Discover the HTTP Port at Runtime How-To

See gh-22898

snicoll committed:
Polish "Harmonize Discover the HTTP Port at Runtime How-To"

See gh-22898

snicoll committed:
Merge pull request #22898 from johncfranco

* pr/22898:
  Polish "Harmonize Discover the HTTP Port at Runtime How-To"
  Harmonize Discover the HTTP Port at Runtime How-To

Closes gh-22898

snicoll committed:
Merge branch '2.2.x' into 2.3.x

Closes gh-22899

snicoll committed:
Merge branch '2.3.x'

Closes gh-22900

snicoll committed:
Polish "Add basic auth support for Prometheus pushgateway"

See gh-22548

snicoll committed:
Merge pull request #22548 from AndrewDi

* pr/22548:
  Polish "Add basic auth support for Prometheus pushgateway"
  Add basic auth support for Prometheus pushgateway

Closes gh-22548

snicoll committed:
Upgrade to Spring Data 2020.0.0-M2

Closes gh-22647

wilkinsona committed:
Simplify build.gradle for spring-boot-starter-data-solr

Closes gh-22903

wilkinsona committed:
Merge branch '2.3.x'

Closes gh-22904

wilkinsona committed:
Upgrade to Gradle 6.6

Closes gh-22905

wilkinsona committed:
Merge branch '2.3.x'

Closes gh-22907

snicoll committed:
Restore handling of certain spring.data.neo4j.* properties

This commit switches the deprecation level of several properties in the
spring.data.neo4j namespace from error and warning. The server URI and
basic authentication can be used in a deprecated fashion to configure
the new neo4j driver.

Closes gh-22653

wilkinsona committed:
Merge pull request #22765 from izeye

* gh-22765:
  Mention configuration customizers for WebTestClient and RestAssured

Closes gh-22765

wilkinsona committed:
Merge branch '2.2.x' into 2.3.x

Closes gh-22910

wilkinsona committed:
Merge branch '2.3.x'

Closes gh-22911

wilkinsona committed:
Improve diagnostics when remote application does not start as expected

See gh-22909

wilkinsona committed:
Merge branch '2.3.x'

See gh-22909

smlc committed:
Fix conditions that lead to auto-configure PushGateway

See gh-22919

snicoll committed:
Upgrade to Spring Data Neumann SR3

Closes gh-22844

snicoll committed:
Merge branch '2.3.x'

philwebb committed:
Fix spring.config.activate.on-profile metadata

Fix the `spring.config.activate.on-profile` entry which was accidentally
entered as `spring.config.import.on-profile`.

See gh-22497

wilkinsona committed:
Fix configuration of environment in Kotlin bootBuildImage examples

Fixes gh-22913

wilkinsona committed:
Merge branch '2.3.x'

Closes gh-22917

wilkinsona committed:
Upgrade to Spring AMQP 2.3.0-M2

Closes gh-22761

jnizet committed:
Fix custom image name example and documentation

There is no `artifactId` property in a gradle Project, and the default
value uses the project name.

See gh-22918

wilkinsona committed:
Merge pull request #22710 from dreis2211

* gh-22710:
  Introduce @ForkedClassPath for testing unmodified class path

Closes gh-22710

wilkinsona committed:
Upgrade to Spring Kafka 2.6.0-M1

Closes gh-22730

wilkinsona committed:
Upgrade to Spring AMQP 2.2.10.RELEASE

Closes gh-22842

wilkinsona committed:
Upgrade to Spring Kafka 2.5.5.RELEASE

Closes gh-22843

wilkinsona committed:
Merge branch '2.3.x'

wilkinsona committed:
Upgrade to Spring Integration 5.4.0-M2

Closes gh-22729

scottfrederick committed:
Merge pull request #22736 from anshlykov

* gh-22736:
  Polish "Add pullPolicy option for image building"
  Add pullPolicy option for image building

Closes gh-22736

scottfrederick committed:
Use Java version wildcard in image building docs

Fixes gh-22916

scottfrederick committed:
Merge branch '2.3.x'

Closes gh-22923

scottfrederick committed:
Fix test for Java version in image building docs

See gh-22916

scottfrederick committed:
Merge branch '2.3.x'

snicoll committed:
Fix default value of server.jetty.accesslog.format

Closes gh-22921

snicoll committed:
Merge branch '2.3.x'

Closes gh-22928

wilkinsona committed:
Upgrade to Spring Batch 4.3.0-M2

Closes gh-22892

snicoll committed:
Upgrade Java 15 version in CI image

Closes gh-22920

snicoll committed:
Polish "Fix conditions that lead to auto-configure PushGateway"

See gh-22919

snicoll committed:
Merge pull request #22919 from smlc

* pr/22919:
  Polish "Fix conditions that lead to auto-configure PushGateway"
  Fix conditions that lead to auto-configure PushGateway

Closes gh-22919

snicoll committed:
Merge branch '2.2.x' into 2.3.x

Closes gh-22930

snicoll committed:
Merge branch '2.3.x'

Closes gh-22931

wilkinsona committed:
Upgrade to Micrometer 1.5.4

Closes gh-22934

wilkinsona committed:
Merge branch '2.3.x'

snicoll committed:
Upgrade to Micrometer 1.5.4

Closes gh-22932

wilkinsona committed:
Upgrade to Flyway 6.5.4

Closes gh-22935

wilkinsona committed:
Upgrade to Micrometer 1.5.4

Closes gh-22936

wilkinsona committed:
Upgrade to Selenium HtmlUnit 2.43.1

Closes gh-22937

wilkinsona committed:
Merge branch '2.3.x'

wilkinsona committed:
Revert "Merge pull request #21060 from dsyer"

This reverts commit 6547ea56f9a6e90ce886c4ba73ca852ccb648bec, reversing
changes made to e9e4a34e622ba3652d4f65e3a1433bed725907fc.

Fixes gh-22039

wilkinsona committed:
Merge branch '2.3.x'

Closes gh-22938

snicoll committed:
Do not configure a ReactiveNeo4jTransactionManager

The current arrangement with Neo4j does not allow us to know for sure
that the user intends to use `@Transactional` for reactive or imperative
operations.

The main reason for that is that the only trigger for the reactive
auto-configuration is the presence of reactor on the classpath given
that the Neo4j driver ships with both styles in the same jar. And
reactor can be on the classpath for a number of reasons that are
unrelated to imperative access (typically `WebClient`).

This commits removes the auto-configuration of the
ReactiveNeo4jAutoConfiguration and makes sure that if the user defines
one, the regular `PlatformTransactionManager` backs off. This is
a temporarily situation until we can provide a better signal that the
user intends to primarily use imperative repositories.

Closes gh-22940

snicoll committed:
Document Spring Data Neo4j 6 support

Closes gh-22652

spring-builds committed:
Next development version (v2.3.4.BUILD-SNAPSHOT)

wilkinsona committed:
Merge branch '2.3.x'

scottfrederick committed:
Fix DataNeo4jTestReactiveIntegrationTests

See gh-22940

philwebb committed:
Polish

philwebb committed:
Rename volumemount -> configtree

Closes gh-22941

dreis2211 committed:
Fix typos in docs

See gh-22945

scottfrederick committed:
Polish "Fix include exception handling in DefaultErrorAttributes"

See gh-22750

maciejwalkowiak committed:
Fix typo `Assert4J` -> `AssertJ`

See gh-22948

wilkinsona committed:
Wait longer for server port and improve diagnostics on timeout

See gh-22909

wilkinsona committed:
Merge branch '2.3.x'

mbhave committed:
Make security auto-configs back off when SecurityFilterChain present

Closes gh-22739

philwebb committed:
Merge pull request #22948 from maciejwalkowiak

* pr/22948:
  Fix typo `Assert4J` -> `AssertJ`

Closes gh-22948

philwebb committed:
Merge branch '2.2.x' into 2.3.x

Closes gh-22960

philwebb committed:
Merge branch '2.3.x'

Closes gh-22961

philwebb committed:
Polish method order

philwebb committed:
Fix and improve SpringApplicationBuilder javadoc

Update SpringApplicationBuilder.properties method javadoc to fix
inaccuracies and apply more consistency.

Closes gh-22962

philwebb committed:
Merge branch '2.2.x' into 2.3.x

Closes gh-22964

philwebb committed:
Merge branch '2.3.x'

Closes gh-22965

philwebb committed:
Add @Interited to @TypeExcludeFilters

Update `@TypeExcludeFilters` so that it is also annotated with
`@Inherited`.

Closes gh-22939

philwebb committed:
Merge branch '2.2.x' into 2.3.x

Closes gh-22966

philwebb committed:
Merge branch '2.3.x'

Closes gh-22967

philwebb committed:
Polish 'Order metrics auto-configurations correctly'

See gh-21134

anshlykov committed:
Consider RecordFilterStrategy in Kafka auto-configuration

See gh-22973

snicoll committed:
Polish

See gh-22945

snicoll committed:
Merge branch '2.3.x'

snicoll committed:
Merge pull request #22945 from dreis2211

* pr/22945:
  Fix typos in docs

Closes gh-22945

snicoll committed:
Polish "Consider RecordFilterStrategy in Kafka auto-configuration"

See gh-22973

snicoll committed:
Merge pull request #22973 from anshlykov

* pr/22973:
  Polish "Consider RecordFilterStrategy in Kafka auto-configuration"
  Consider RecordFilterStrategy in Kafka auto-configuration

Closes gh-22973

snicoll committed:
Remove trailing whitespaces

snicoll committed:
Merge pull request #22918 from jnizet

* pr/22918:
  Fix custom image name example and documentation

Closes gh-22918

snicoll committed:
Merge branch '2.3.x'

Closes gh-22980

philwebb committed:
Merge pull request #22750 from Lopfest

* pr/22750:
  Polish "Fix include exception handling in DefaultErrorAttributes"
  Fix include exception handling in DefaultErrorAttributes

Closes gh-22750

philwebb committed:
Merge branch '2.3.x'

Closes gh-22984

philwebb committed:
Merge pull request #21134 from bendiscz

* pr/21134:
  Polish 'Order metrics auto-configurations correctly'
  Order metrics auto-configurations correctly

Closes gh-21134

philwebb committed:
Merge branch '2.2.x' into 2.3.x

Closes gh-22989

philwebb committed:
Merge branch '2.3.x'

Closes gh-22990

philwebb committed:
Rename CompositeMeterRegistryAutoConfiguration

Rename `CompositeMeterRegistryAutoConfiguration` to
`MeterRegistryAutoConfiguration` since it can also create non-composite
registries.

Closes gh-22988

snicoll committed:
Rename neo4j.version to neo4j-java-driver.version

Closes gh-22952

philwebb committed:
Attempt to fix memory leak in JarFile class

Create a new `JarFileWrapper` class so that we can wrap and existing
`JarFile` and offer a version that can be safely closed.

Prior to this commit, we provided wrapper functionality in the `JarFile`
class itself. Unfortunately, because we override `close` and also create
a lot of wrappers this caused memory issues when running on Java 11.

With Java 11 `java.util.zip.ZipFile` class uses `FinalizableResource`
for any implementation that overrides `close()`. This means that any
wrapper classes will not be garbage collected until the JVM finalizer
thread runs.

Closes gh-22991

philwebb committed:
Merge branch '2.2.x' into 2.3.x

Closes gh-22998

philwebb committed:
Merge branch '2.3.x'

Closes gh-22999

scottfrederick committed:
Fail on layertools extract with launch script

This commit adds a check to the `layertools extract` command to
ensure that the jar file being processed is readable and has a
valid directory.

Fixes gh-22993

scottfrederick committed:
Merge branch '2.3.x'

Closes gh-23000

philwebb committed:
Refine AbstractJarFile method visibility

Refine method visibility in an attempt to fix test issues on Java 11+.

See gh-22991

philwebb committed:
Update copyright year of changed files

philwebb committed:
Merge branch '2.2.x' into 2.3.x

philwebb committed:
Merge branch '2.2.x' into 2.3.x

philwebb committed:
Merge branch '2.3.x'

philwebb committed:
Update copyright year of changed files

philwebb committed:
Merge branch '2.3.x'

philwebb committed:
Update copyright year of changed files

philwebb committed:
Remove Mockito from JarFileWrapperTests

Remove Mockto from JarFileWrapperTests since it seems to be failing on
later versions of Java.

See gh-22991

philwebb committed:
Merge branch '2.2.x' into 2.3.x

philwebb committed:
Merge branch '2.3.x'

philwebb committed:
Ensure LayoutFactory is not passed an empty file

Update `Repackager` to ensure that `getLayout` is called before we
backup the source file. This restores earlier behavior that some
custom `ModuleFactory` implementations were relying on.

Closes gh-22995

philwebb committed:
Merge branch '2.3.x'

Closes gh-23001

dreis2211 committed:
Mention correct JUnit 5 annotations in Kotlin testing section

See gh-23002

snicoll committed:
Merge pull request #23002 from dreis2211

* pr/23002:
  Mention correct JUnit 5 annotations in Kotlin testing section

Closes gh-23002

snicoll committed:
Merge branch '2.2.x' into 2.3.x

Closes gh-23004

snicoll committed:
Merge branch '2.3.x'

Closes gh-23005

emerkle826 committed:
Add Cassandra driver Actuator metrics

emerkle826 committed:
[fixup] update to Datastax driver 4.9.0

emerkle826 committed:
[fixup] Use Testcontainers wait strategy
